### PR TITLE
[SC-241] Refactor Base/Kink Rate Target IRMs

### PR DESCRIPTION
**Pull request by hexonaut**
To reduce code duplication these IRMs will inherit from the custom variable rate IRM which removes stable borrow logic.

### CODE REVIEW

**Comment by hexonaut** [comment]
Coverage seems broken...?

**Review by barrutko — APPROVED** [review]
✅ In `RateTargetBaseInterestRateStrategy` the behaviour of `getBaseVariableBorrowRate()` is correct
✅ In `RateTargetKinkInterestRateStrategy` the behaviour of `getVariableRateSlope1()` is correct
✅ `VariableBorrowInterestRateStrategy` refactoring makes sense
✅ I didn't find any style-related issues, assuming that we keep Aave style on purpose in some places

IMO this is good to be merged

File: src/RateTargetBaseInterestRateStrategy.sol [hunk 1]
```diff
@@ -1,264 +1,55 @@
...
-import {IERC20} from 'aave-v3-core/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
-import {WadRayMath} from 'aave-v3-core/contracts/protocol/libraries/math/WadRayMath.sol';
-import {PercentageMath} from 'aave-v3-core/contracts/protocol/libraries/math/PercentageMath.sol';
-import {DataTypes} from 'aave-v3-core/contracts/protocol/libraries/types/DataTypes.sol';
-import {Errors} from 'aave-v3-core/contracts/protocol/libraries/helpers/Errors.sol';
-import {IDefaultInterestRateStrategy} from 'aave-v3-core/contracts/interfaces/IDefaultInterestRateStrategy.sol';
-import {IReserveInterestRateStrategy} from 'aave-v3-core/contracts/interfaces/IReserveInterestRateStrategy.sol';
-import {IPoolAddressesProvider} from 'aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol';
-
 import {IRateSource} from './interfaces/IRateSource.sol';
```
Comment (lucas-manuel) [inline src/RateTargetBaseInterestRateStrategy.sol:5]:
```suggestion
import { IRateSource } from './interfaces/IRateSource.sol';
```

File: src/RateTargetKinkInterestRateStrategy.sol [hunk 2]
```diff
@@ -1,270 +1,58 @@
...
-import {IERC20} from 'aave-v3-core/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
-import {WadRayMath} from 'aave-v3-core/contracts/protocol/libraries/math/WadRayMath.sol';
-import {PercentageMath} from 'aave-v3-core/contracts/protocol/libraries/math/PercentageMath.sol';
-import {DataTypes} from 'aave-v3-core/contracts/protocol/libraries/types/DataTypes.sol';
-import {Errors} from 'aave-v3-core/contracts/protocol/libraries/helpers/Errors.sol';
-import {IDefaultInterestRateStrategy} from 'aave-v3-core/contracts/interfaces/IDefaultInterestRateStrategy.sol';
-import {IReserveInterestRateStrategy} from 'aave-v3-core/contracts/interfaces/IReserveInterestRateStrategy.sol';
-import {IPoolAddressesProvider} from 'aave-v3-core/contracts/interfaces/IPoolAddressesProvider.sol';
-
 import {IRateSource} from './interfaces/IRateSource.sol';
```
Comment (lucas-manuel) [inline src/RateTargetKinkInterestRateStrategy.sol:5]:
```suggestion
import { IRateSource } from './interfaces/IRateSource.sol';
```

File: src/RateTargetKinkInterestRateStrategy.sol [hunk 3]
```diff
@@ -1,270 +1,58 @@
...
+    int256 internal immutable _variableRateSlope1Spread;
+
+    /**
+     * @dev Constructor.
+     * @param provider                 The address of the PoolAddressesProvider contract
+     * @param rateSource               The address of the rate source contract
+     * @param optimalUsageRatio        The optimal usage ratio
+     * @param baseVariableBorrowRate   The base variable borrow rate
+     * @param variableRateSlope1Spread The spread between the rate source and the desired target kink rate
+     *                                 Note: this discounts the base variable borrow rate
```
Comment (lucas-manuel) [inline src/RateTargetKinkInterestRateStrategy.sol]: This

File: src/RateTargetBaseInterestRateStrategy.sol [hunk 4]
```diff
@@ -1,264 +1,55 @@
...
+contract RateTargetBaseInterestRateStrategy is VariableBorrowInterestRateStrategy {
+
+    IRateSource public immutable RATE_SOURCE;
+
+    // Base variable borrow rate spread when usage rate = 0. Expressed in ray
+    uint256 internal immutable _baseVariableBorrowRateSpread;
+
+    /**
+     * @dev Constructor.
+     * @param provider                     The address of the PoolAddressesProvider contract
```
Comment (lucas-manuel) [inline src/RateTargetBaseInterestRateStrategy.sol]: Periods at end of all natspec

File: src/VariableBorrowInterestRateStrategy.sol [hunk 5]
```diff
@@ -21,22 +21,16 @@ import { IPoolAddressesProvider }       from 'aave-v3-core/contracts/interfaces/
...
-    uint256 public constant MAX_EXCESS_STABLE_TO_TOTAL_DEBT_RATIO = WadRayMath.RAY;
+    uint256 public override immutable OPTIMAL_USAGE_RATIO;
+    uint256 public override constant  OPTIMAL_STABLE_TO_TOTAL_DEBT_RATIO = 0;
+    uint256 public override immutable MAX_EXCESS_USAGE_RATIO;
+    uint256 public override constant  MAX_EXCESS_STABLE_TO_TOTAL_DEBT_RATIO = WadRayMath.RAY;
 
-    IPoolAddressesProvider public immutable ADDRESSES_PROVIDER;
+    IPoolAddressesProvider public override immutable ADDRESSES_PROVIDER;
 
     // Base variable borrow rate when usage rate = 0. Expressed in ray
```
Comment (lucas-manuel) [inline src/VariableBorrowInterestRateStrategy.sol:35]: Expressed in ray.

**Review by lucas-manuel — CHANGES_REQUESTED** [review]
Small changes then LGTM, ran fuzz tests with 100k runs and got the same issue with the rounding and that fix fixes it so lgtm after this.

File: src/RateTargetBaseInterestRateStrategy.sol (hunk 1 above)
Comment (hexonaut) [reply src/RateTargetBaseInterestRateStrategy.sol:5]: fixed

File: src/RateTargetKinkInterestRateStrategy.sol (hunk 2 above)
Comment (hexonaut) [reply src/RateTargetKinkInterestRateStrategy.sol:5]: fixed

File: src/RateTargetKinkInterestRateStrategy.sol (hunk 3 above)
Comment (hexonaut) [reply src/RateTargetKinkInterestRateStrategy.sol]: fixed

File: src/RateTargetBaseInterestRateStrategy.sol (hunk 4 above)
Comment (hexonaut) [reply src/RateTargetBaseInterestRateStrategy.sol]: fixed

File: src/VariableBorrowInterestRateStrategy.sol (hunk 5 above)
Comment (hexonaut) [reply src/VariableBorrowInterestRateStrategy.sol:35]: fixed

**Review by barrutko — APPROVED** [review]

**Review by lucas-manuel — APPROVED** [review]

**Pull request merged**